File: small-doc-ie-bench/src/docie_bench/inngest/studio_api/datasets.py
"""Registered-dataset listing + validation routes."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query

router = APIRouter()


@router.get("/datasets")
async def list_datasets() -> list[dict[str, Any]]:
    """Registered benchmark datasets (``data/datasets.yaml``) — what a "Dataset"
    field in the Studio can actually reference, so the Benchmark form can offer
    a picker instead of a free-text guess.

    Each entry is a name (the ``dataset`` value ``/benchmark`` and
    ``run_benchmark`` resolve), its latest version, all known versions, and the
    latest version's document count. Missing registry file degrades to an
    empty list (``load_registry``'s own contract), never a 500.
    """
    from docie_bench.benchmark.registry import DEFAULT_REGISTRY_PATH, load_registry

    registry = load_registry(DEFAULT_REGISTRY_PATH)
    return [
        {
            "name": name,
            "description": record.description,
            "latest": record.latest,
            "versions": sorted(record.versions),
            "documents": (
                record.versions[record.latest].statistics.get("documents")
                if record.latest and record.latest in record.versions
                else None
            ),
        }
        for name, record in sorted(registry.datasets.items())
    ]


@router.post("/datasets/{name}/validate")
async def validate_dataset_version(
    name: str,
    version: str | None = None,
    near_duplicate_threshold: float = Query(default=0.92, ge=0.0, le=1.0),
) -> dict[str, Any]:
    """Run the CLI's dataset validation (`docie-bench dataset validate`/`inspect`) against
    an already-registered dataset version, reachable from the Studio.

    Resolves the reference through the shared `data/datasets.yaml` registry (this Studio
    API already reads that file directly for `GET /datasets`, so this isn't a new
    filesystem assumption). Unlike either CLI command, this deliberately does NOT use
    `resolve_dataset`'s own `verify_hash` gate (which raises and stops before validation
    runs) -- passing the registry's recorded hash as `expected_hash` instead lets a pure
    drift case (files changed on disk since registration, nothing else wrong) surface
    inside the SAME report as leakage/statistics, rather than a bare error.

    This does NOT mean drift and structural problems always merge into one report:
    `validate_dataset` (registry.py) early-returns as soon as any duplicate-doc_id /
    missing-file / unsupported-suffix error exists, before it ever reaches the hash
    comparison -- same short-circuit it's always had. If a dataset has both a structural
    error and real drift, only the structural error is reported; the hash is never
    checked in that case. See
    test_validate_endpoint_reports_corruption_introduced_after_registration.

    `resolve_dataset` itself still unconditionally hashes every referenced file (even
    with verify_hash=False) -- so a manifest referencing an ENTIRELY missing file raises
    OSError there, before validate_dataset's own graceful per-item existence check ever
    runs (the CLI's `dataset_validate` has this identical gap, see cli.py:363's shared
    `except (OSError, ValueError)`). Surfaced as 422 (the registered dataset itself is
    broken) rather than a raw 500.
    """
    from docie_bench.benchmark.registry import (
        DEFAULT_REGISTRY_PATH,
        load_registry,
        resolve_dataset,
        validate_dataset,
    )

    reference = f"{name}@{version}" if version else name
    try:
        resolved = resolve_dataset(
            reference, registry_path=DEFAULT_REGISTRY_PATH, verify_hash=False
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except OSError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    expected_hash = None
    if resolved.version is not None:
        record = load_registry(DEFAULT_REGISTRY_PATH).datasets.get(name)
        entry = record.versions.get(resolved.version) if record else None
        expected_hash = entry.dataset_hash if entry else None
    report = validate_dataset(
        resolved.manifest_path,
        near_duplicate_threshold=near_duplicate_threshold,
        expected_hash=expected_hash,
    )
    return {
        # No manifest_path here: an absolute server-side filesystem path, unlike every
        # other artifact this Studio API exposes (see download_artifact's own explicit
        # principle -- resolved via id -> DB row -> shared blob store, never a
        # worker-local path). The CLI's dataset_inspect prints it because it IS the
        # local tool; a network API shouldn't.
        "reference": resolved.reference,
        "version": resolved.version,
        **report,
    }

File: small-doc-ie-bench/frontend/lib/api/datasets.ts
// Registered-dataset listing + validation (duplicate doc_id, missing files,
// cross-split leakage, stats).

import { request } from "./core";

/** A registered benchmark dataset (GET /v1/studio/datasets) — the values a
 * Benchmark run's `dataset` field can actually reference. */
export interface DatasetSummary {
  name: string;
  description: string | null;
  latest: string | null;
  versions: string[];
  documents: number | null;
}

/** Registered datasets, so the Benchmark form can offer a picker instead of a
 * free-text guess at a name that may not be registered. */
export function listDatasets(): Promise<DatasetSummary[]> {
  return request<DatasetSummary[]>("/v1/studio/datasets");
}

interface DatasetLeakagePair {
  left_doc_id: string;
  left_split: string;
  right_doc_id: string;
  right_split: string;
  similarity?: number;
}

export interface DatasetLeakageReport {
  near_duplicate_threshold: number;
  exact_duplicates: DatasetLeakagePair[];
  near_duplicates: DatasetLeakagePair[];
  leakage_pairs: number;
}

export interface DatasetStatistics {
  documents: number;
  total_bytes: number;
  schemas: Record<string, number>;
  languages: Record<string, number>;
  splits: Record<string, number>;
  ground_truth_fields: Record<string, number>;
  labeled_documents: number;
}

// dataset_hash/statistics/leakage are ABSENT (not merely empty) when validate_dataset
// short-circuits on a structural error (duplicate doc_id, missing file) before it ever
// reaches the hash/leakage/stats computation -- see build_comparison_payload's sibling
// docstring in studio_api.py for the same "one report, but not every field always
// present" contract.
export interface DatasetValidationReport {
  reference: string;
  version: string | null;
  valid: boolean;
  errors: string[];
  warnings: string[];
  dataset_hash?: string;
  statistics?: DatasetStatistics;
  leakage?: DatasetLeakageReport;
}

/** Run the CLI's dataset validate/inspect checks against an already-registered
 * dataset version (duplicate doc_id, missing files, cross-split leakage, stats). */
export function validateDataset(
  name: string,
  version?: string,
): Promise<DatasetValidationReport> {
  const params = new URLSearchParams();
  if (version) params.set("version", version);
  const qs = params.toString();
  return request<DatasetValidationReport>(
    `/v1/studio/datasets/${encodeURIComponent(name)}/validate${qs ? `?${qs}` : ""}`,
    { method: "POST" },
  );
}
